include/deal.II/matrix_free/evaluation_selector.h

```diff
@@ -0,0 +1,573 @@
+    }
+
+    static inline void integrate (const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<Number> > &shape_info,
+                                  VectorizedArray<Number> *values_dofs_actual[],
+                                  VectorizedArray<Number> *values_quad[],
+                                  VectorizedArray<Number> *gradients_quad[][dim],
+                                  VectorizedArray<Number> *scratch_data,
+                                  const bool               integrate_values,
+                                  const bool               integrate_gradients)
+    {
+      internal::FEEvaluationImpl<internal::MatrixFreeFunctions::tensor_general,
+               dim, -1, 0, n_components, Number>
+               ::integrate(shape_info, values_dofs_actual, values_quad,
+                           gradients_quad, scratch_data,
+                           integrate_values, integrate_gradients);
+    }
+  };
+
+
+  /**
+   * This class implements the actual choice of the template specialization.
+   */
+  template<int dim, int n_components, typename Number,
+           int DEPTH=0, int degree=0, int n_q_points_1d=0, class Enable = void>
+  struct Factory : Default<dim, n_components, Number> {};
+
+  /**
+   * This specialization sets the maximal fe_degree for
+   * which we want to determine the correct template parameters based at runtime.
+   */
+  template<int n_q_points_1d, int dim, int n_components, typename Number>
+  struct Factory<dim, n_components, Number, 0, 10, n_q_points_1d> : Default<dim, n_components, Number> {};
+
+  /**
+   * This specialization sets the maximal number of n_q_points_1d for
+   * which we want to determine the correct template parameters based at runtime.
+   */
+  template<int degree, int n_q_points_1d, int dim, int n_components, typename Number>
+  struct Factory<dim, n_components, Number, 1, degree, n_q_points_1d,
+    typename std::enable_if<n_q_points_1d==degree+3>::type> : Default<dim, n_components, Number> {};
+
+  /**
+   * This class chooses the correct template degree.
+   */
+  template<int degree, int n_q_points_1d, int dim, int n_components, typename Number>
+  struct Factory<dim, n_components, Number, 0, degree, n_q_points_1d>
+  {
+    static inline void evaluate (
+      const internal::MatrixFreeFunctions::ShapeInfo<VectorizedArray<Number> > &shape_info,
+      VectorizedArray<Number> *values_dofs_actual[],
+      VectorizedArray<Number> *values_quad[],
+      VectorizedArray<Number> *gradients_quad[][dim],
+      VectorizedArray<Number> *hessians_quad[][(dim*(dim+1))/2],
+      VectorizedArray<Number> *scratch_data,
+      const bool               evaluate_values,
+      const bool               evaluate_gradients,
+      const bool               evaluate_hessians)
+    {
+      const unsigned int runtime_degree = shape_info.fe_degree;
+      constexpr unsigned int nonnegative_start_n_q_points = (degree>0)?degree-1:0;
```

Comment:
Can you make this similar to the code in `integrate`, i.e., start with `degree+1`? Then this should be good for merging (we can leave it at 0 to 10 if nobody else has an opinion on this).